File: SpiralTraverseMatrix.java
// Time Complexity : O(N.M) N=rows M=colums
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
public class SpiralTraverseMatrix {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> result = new ArrayList<>();
        int rows = matrix.length;
        int columns = matrix[0].length;
        int up = 0;
        int left = 0;
        int right = columns - 1;
        int down = rows - 1;

        while (result.size() < rows * columns) {
            for (int col = left; col <= right; col++) {
                result.add(matrix[up][col]);
            }
            for (int row = up + 1; row <= down; row++) {
                result.add(matrix[row][right]);
            }
            if (up != down) {
                for (int col = right - 1; col >= left; col--) {
                    result.add(matrix[down][col]);
                }
            }
            if (left != right) {
                // Traverse upwards.
                for (int row = down - 1; row > up; row--) {
                    result.add(matrix[row][left]);
                }
            }
            left++;
            right--;
            up++;
            down--;
        }

        return result
    }
}

File: spiral_traverse_matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m=len(matrix)
        n=len(matrix[0])
        l=0
        r=n-1
        t=0
        d=m-1
        
        res=[]
        while l<=r and t<=d:
            for i in range(l,r+1):
                res.append(matrix[t][i])
                print(str(i)+" t= "+str(t)+" ",end=" ")
            t+=1
            print()
            for i in range(t,d+1):
                res.append(matrix[i][r])
                print(str(i)+" r= "+str(r)+" ",end=" ")
            r-=1
            print()

            if t<=d:
                for i in range(r,l-1,-1):
                    res.append(matrix[d][i])
                    print(str(i)+" d= "+str(d)+" ",end=" ")
                d-=1
            print()
            if l<=r:
                for i in range(d,t-1,-1):
                    res.append(matrix[i][l])
                    print(str(i)+" l= "+str(l)+" ",end=" ")
                l+=1
            print()
        print(res)

        return res
        